api/agent/resource.go

```diff
@@ -284,26 +234,32 @@ func (a *resourceTracker) GetResourceToken(ctx context.Context, memory uint64, c
 	go func() {
 		defer span.End()
 		defer cancel()
+
+		// If context times out or canceled, we return CapacityFull
+		var t ResourceToken
+		t = &resourceToken{err: CapacityFull}
+
 		c.L.Lock()
 
 		isWaiting = true
-		for !a.isResourceAvailableLocked(memory, cpuQuota, isAsync) && ctx.Err() == nil {
+		for {
+			// Order is important. We do want to pass a token even if ctx is expired
+			if a.isResourceAvailableLocked(memory, cpuQuota, isAsync) {
+				t = a.allocResourcesLocked(memory, cpuQuota, isAsync)
+				break
+			}
+			if ctx.Err() != nil {
+				break
+			}
 			c.Wait()
 		}
 		isWaiting = false
 
-		if ctx.Err() != nil {
-			c.L.Unlock()
-			return
-		}
-
-		t := a.allocResourcesLocked(memory, cpuQuota, isAsync)
 		c.L.Unlock()
 
 		select {
 		case ch <- t:
-		case <-ctx.Done():
```

Comment:
there is a race here where the caller hasn't entered the select yet but this goroutine reaches here. i forget if we added this because we ran into that or not.

Reply:
Hmm, with current agent.go usage, this is OK, but this is fragile... If I add back the ctx, then that's also problematic. I'll figure out and submit a fix..